### ISSUE TITLE
Consul intermittent failure with a server with dual NIC

### ISSUE THREAD

**Issue by ckim0419** [hashicorp/consul#1015]
We have Consul setup on a host with dual NIC, and having Consul listening on both port. From time to time, when consul restart, we get following error messages:

```
consul_1 |     2015/06/09 17:32:55 [ERR] consul: failed to wait for barrier: leadership lost while committing log
consul_1 |     2015/06/09 17:32:55 [INFO] raft: pipelining replication to peer 192.168.50.4:8300
consul_1 |     2015/06/09 17:32:55 [INFO] raft: aborting pipeline replication to peer 192.168.50.4:8300
consul_1 |     2015/06/09 17:32:55 [INFO] consul: cluster leadership lost
consul_1 |     2015/06/09 17:32:55 [INFO] consul: New leader elected: ubuntu-14
consul_1 |     2015/06/09 17:32:56 [ERR] agent: failed to sync remote state: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc
error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc erro
r: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: r
pc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc e
rror: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error
: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rp
c error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc error: rpc er
```

**Comment by ryanuber** [comment]
We have seen this before but haven't got a solid repro case yet. Thanks for reporting it. Can you make a gist with the un-truncated logs? Also, if you could start Consul with `-log-level=debug`, that might give us a little more context and help get this solved.

**Comment by ckim0419** [comment]
Thanks for the prompt response. Here it is.

https://gist.github.com/ckim0419/6d030c6201b0c1e2d4ac

Here's how to I reproduce:
1. Setup a server with two NIC and assign IPs to both.
2. Run Consul, server mode with bootstrap enabled.
3. Wait until bootstrap finishes.
4. Restart the entire machine.
5. Start Consul.

**Comment by robinpercy** [comment]
@ckim0419 Looks like this was fixed in 0.5.1 https://github.com/hashicorp/consul/issues/611

**Comment by KIVagant** [comment]
Hello.
I am not an expert in consul. But I think, that I have similar issue and I find how to fix this in my case.
```
consul -v
Consul v0.5.2
Consul Protocol: 2 (Understands back to: 1)
```

My config-file looks like:

```
{
  "ui_dir": "/opt/consul/dist",
  "disable_remote_exec": true,
  "domain": "consul.",
  "data_dir": "/opt/consul/data",
  "log_level": "INFO",
  "enable_syslog": true,
  "node_name": "labs_node",
  "server": true,
  "client_addr": "127.0.0.1",
  "bind_addr": "0.0.0.0",           ← TL;DR : here is a problem
  "datacenter": "labs",
  "bootstrap": false,
  "bootstrap_expect": 1,
  "ports": {
    "dns": 8600,
    "http": 8500,
    "https": -1,
    "rpc": 8400,
    "serf_lan": 8301,
    "serf_wan": 8302,
    "server": 8300
  },
  "rejoin_after_leave": true,
  "leave_on_terminate": false
}
```

And ```consul monitor``` shows me this:

```
2015/12/14 14:28:31 [WARN] raft: Heartbeat timeout reached, starting election
2015/12/14 14:28:31 [INFO] raft: Node at 192.168.0.218:8300 [Candidate] entering Candidate state
2015/12/14 14:28:31 [DEBUG] raft-net: 192.168.0.218:8300 accepted connection from: 10.0.3.1:51540
2015/12/14 14:28:31 [DEBUG] raft: Votes needed: 2
2015/12/14 14:28:31 [INFO] raft: Duplicate RequestVote for same term: 1116575
2015/12/14 14:28:31 [WARN] raft: Duplicate RequestVote from candidate: 192.168.0.218:8300
2015/12/14 14:28:31 [DEBUG] raft: Vote granted. Tally: 1
2015/12/14 14:28:31 [WARN] raft: Remote peer 10.0.3.1:8300 does not have local node 192.168.0.218:8300 as a peer
2015/12/14 14:28:31 [DEBUG] raft: Vote granted. Tally: 2
2015/12/14 14:28:31 [INFO] raft: Election won. Tally: 2
2015/12/14 14:28:31 [INFO] raft: Node at 192.168.0.218:8300 [Leader] entering Leader state
2015/12/14 14:28:31 [INFO] consul: cluster leadership acquired
2015/12/14 14:28:31 [INFO] consul: New leader elected: labs_node
2015/12/14 14:28:31 [INFO] raft: Node at 192.168.0.218:8300 [Follower] entering Follower state
2015/12/14 14:28:31 [INFO] consul: cluster leadership lost
2015/12/14 14:28:31 [ERR] consul: failed to wait for barrier: node is not the leader
2015/12/14 14:28:31 [INFO] raft: pipelining replication to peer 10.0.3.1:8300
2015/12/14 14:28:31 [INFO] raft: aborting pipeline replication to peer 10.0.3.1:8300
2015/12/14 14:28:31 [WARN] raft: Clearing log suffix from 1401609 to 1401609
2015/12/14 14:28:32 [WARN] serf: Event queue depth: 223
2015/12/14 14:28:32 [ERR] agent: failed to sync remote state: rpc error: rpc error: 
... ... ...
 rpc error: No cluster leader
2015/12/14 14:28:33 [WARN] serf: Event queue depth: 223
2015/12/14 14:28:33 [WARN] raft: Heartbeat timeout reached, starting election
2015/12/14 14:28:33 [INFO] raft: Node at 192.168.0.218:8300 [Candidate] entering Candidate state
2015/12/14 14:28:33 [DEBUG] raft-net: 192.168.0.218:8300 accepted connection from: 10.0.3.1:51542
```

Looks like consul uses the secondary network interface with IP *10.0.3.1*.

I changed bind-address in config to bind_addr": "192.168.0.218", but the problem does not gone.
Then I executed ```consul leave``` and restarted the service.

And now ```consul monitor``` shows me this:

```
2015/12/14 14:41:05 [INFO] consul.fsm: snapshot created in 172.628µs
2015/12/14 14:41:05 [INFO] raft: Starting snapshot up to 1401694
2015/12/14 14:41:05 [INFO] snapshot: Creating new snapshot at /opt/consul/data/raft/snapshots/1116923-1401694-1450096865278.tmp
2015/12/14 14:41:05 [INFO] snapshot: reaping snapshot /opt/consul/data/raft/snapshots/6-16387-1447554581844
2015/12/14 14:41:05 [INFO] raft: Compacting logs from 14340 to 1391454
2015/12/14 14:41:06 [INFO] raft: Snapshot to 1401694 complete
2015/12/14 14:47:26 [INFO] agent.rpc: Accepted client: 127.0.0.1:49634
```

The problem was solved.

UP: after this fix, I changed bind-address back to 0.0.0.0 and in spite of this all is works fine.

**Comment by slackpad** [comment]
Closing per https://github.com/hashicorp/consul/issues/1015#issuecomment-121467270 and we've done a ton of work since on Raft and Consul in general and haven't seen any recent reports of this.
